test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp

```diff
@@ -346,6 +346,55 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
   // Communicators might be aborted here, further operations would fail.
 }
 
+// Extend the nested class outside the parent class
+class TestDebugInfoCallbackStorer
+    : public c10d::ProcessGroupNCCL::DebugInfoCallbackStorer {
+ public:
+  TestDebugInfoCallbackStorer() : DebugInfoCallbackStorer("") {
+    temp_ = std::tmpfile();
+    TORCH_CHECK(temp_ != nullptr, "Failed to open temporary file");
+  }
+
+  ~TestDebugInfoCallbackStorer() override {
+    std::fclose(temp_);
+  }
+
+  void storeTraceInfoStorer(int rank, const std::string& ncclTrace) override {
+    tracesToStorer_.assign(ncclTrace.begin(), ncclTrace.end());
+    std::fputs(ncclTrace.data(), temp_);
+    std::fflush(temp_);
+    // Reset to the beginning of the file.
+    std::rewind(temp_);
+    std::array<char, 256> buffer;
+    if (std::fgets(buffer.data(), ncclTrace.size(), temp_) != nullptr) {
+      tracesFromStorer_.assign(
+          buffer.begin(), buffer.begin() + ncclTrace.size());
+      // One thing which can not be explained is that, without this line of log,
+      // the last element of `tracesFromStorer_` will be different from
+      // `tracesToStorer_`. Since using TmpFile is just for testing purpose,
+      // this is probably not a big deal.
```

Comment:
better explain it, is it due to our thread exiting before it flushes its file? do we need to flush again before we return?

Reply:
Actually I don't think it is due to thread exiting because we do the write right after read. And we are reading here, we don't need to call flush, right? The weird thing is that if we comment the line of LOG(INFO), the comparison of two vector will be different, which I cannot reason how come it is like that.

Reply:
When using stream and filesystem, this issue is gone..